File: packages/sdks/client-sdks/gatrix-python-sdk/gatrix/flag_proxy.py
"""FlagProxy - Thin shell that delegates ALL logic to VariationProvider.

Architecture per CLIENT_SDK_SPEC:
- Holds only flag_name + force_realtime + client reference.
- ALL property reads and variation methods delegate to the client.
- No deep copy of flag data - always reads live state from FeaturesClient cache.
- is_realtime property indicates the proxy's operational mode.
- Client is always present (never None).
"""
from __future__ import annotations

from typing import Any, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from gatrix.variation_provider import VariationProvider


class FlagProxy:
    """Thin shell around a feature flag that delegates all reads to the VariationProvider."""

    __slots__ = ('_client', '_flag_name', '_force_realtime')

    def __init__(
        self,
        client: VariationProvider,
        flag_name: str,
        force_realtime: bool = False,
    ) -> None:
        self._client = client
        self._flag_name = flag_name or ''
        self._force_realtime = force_realtime

    # ==================== Properties ====================

    @property
    def name(self) -> str:
        return self._flag_name

    @property
    def is_realtime(self) -> bool:
        """Whether this proxy was created in realtime mode."""
        return self._force_realtime

    @property
    def exists(self) -> bool:
        """Whether the flag exists in the current cache."""
        return self._client.has_flag_internal(self._flag_name, self._force_realtime)

    @property
    def enabled(self) -> bool:
        """Check if flag is enabled. Delegates to client for metrics tracking."""
        return self._client.is_enabled_internal(self._flag_name, self._force_realtime)

    @property
    def variant(self):
        return self._client.get_variant_internal(self._flag_name, self._force_realtime)

    @property
    def value_type(self) -> str:
        return self._client.get_value_type_internal(self._flag_name, self._force_realtime)

    @property
    def version(self) -> int:
        return self._client.get_version_internal(self._flag_name, self._force_realtime)

    @property
    def impression_data(self) -> bool:
        return self._client.get_impression_data_internal(self._flag_name, self._force_realtime)

    @property
    def raw(self):
        return self._client.get_raw_flag_internal(self._flag_name, self._force_realtime)

    @property
    def reason(self) -> Optional[str]:
        return self._client.get_reason_internal(self._flag_name, self._force_realtime)

    # ==================== Variation Methods ====================
    # All methods delegate to client's internal methods.
    # FlagProxy is a convenience shell - no own logic.

    def variation(self, fallback_value: str) -> str:
        return self._client.variation_internal(self._flag_name, fallback_value, self._force_realtime)

    def bool_variation(self, fallback_value: bool) -> bool:
        return self._client.bool_variation_internal(self._flag_name, fallback_value, self._force_realtime)

    def string_variation(self, fallback_value: str) -> str:
        return self._client.string_variation_internal(self._flag_name, fallback_value, self._force_realtime)

    def int_variation(self, fallback_value: int) -> int:
        return self._client.int_variation_internal(self._flag_name, fallback_value, self._force_realtime)

    def float_variation(self, fallback_value: float) -> float:
        return self._client.float_variation_internal(self._flag_name, fallback_value, self._force_realtime)

    def json_variation(self, fallback_value: Any) -> Any:
        return self._client.json_variation_internal(self._flag_name, fallback_value, self._force_realtime)

    # ==================== Variation Details ====================

    def bool_variation_details(self, fallback_value: bool):
        return self._client.bool_variation_details_internal(self._flag_name, fallback_value, self._force_realtime)

    def string_variation_details(self, fallback_value: str):
        return self._client.string_variation_details_internal(self._flag_name, fallback_value, self._force_realtime)

    def int_variation_details(self, fallback_value: int):
        return self._client.int_variation_details_internal(self._flag_name, fallback_value, self._force_realtime)

    def float_variation_details(self, fallback_value: float):
        return self._client.float_variation_details_internal(self._flag_name, fallback_value, self._force_realtime)

    def json_variation_details(self, fallback_value: Any):
        return self._client.json_variation_details_internal(self._flag_name, fallback_value, self._force_realtime)

    # ==================== Strict Variation Methods (OrThrow) ====================

    def bool_variation_or_throw(self) -> bool:
        return self._client.bool_variation_or_throw_internal(self._flag_name, self._force_realtime)

    def string_variation_or_throw(self) -> str:
        return self._client.string_variation_or_throw_internal(self._flag_name, self._force_realtime)

    def int_variation_or_throw(self) -> int:
        return self._client.int_variation_or_throw_internal(self._flag_name, self._force_realtime)

    def float_variation_or_throw(self) -> float:
        return self._client.float_variation_or_throw_internal(self._flag_name, self._force_realtime)

    def json_variation_or_throw(self) -> Any:
        return self._client.json_variation_or_throw_internal(self._flag_name, self._force_realtime)

File: packages/sdks/client-sdks/gatrix-unity-sdk/src/Assets/Plugins/Gatrix/Runtime/FlagProxy.cs
// FlagProxy - Thin shell that delegates ALL logic to VariationProvider.
//
// Architecture per CLIENT_SDK_SPEC:
// - Holds only flagName + forceRealtime + client reference.
// - ALL property reads and variation methods delegate to the client.
// - No deep copy of flag data — always reads live state from FeaturesClient cache.
// - isRealtime property indicates the proxy's operational mode.
// - Client is always present (never null).

using System.Collections.Generic;

namespace Gatrix.Unity.SDK
{
    /// <summary>
    /// Convenience wrapper for accessing flag values.
    /// Delegates all logic to FeaturesClient via IVariationProvider.
    /// No internal flag copy — always reads from the client's cache.
    /// </summary>
    public class FlagProxy
    {
        private readonly IVariationProvider _client;
        private readonly string _flagName;
        private readonly bool _forceRealtime;

        public FlagProxy(IVariationProvider client, string flagName, bool forceRealtime = false)
        {
            _client = client;
            _flagName = flagName ?? "";
            _forceRealtime = forceRealtime;
        }

        #region Properties

        public string Name => _flagName;

        /// <summary>Whether this proxy was created in realtime mode.</summary>
        public bool IsRealtime => _forceRealtime;

        /// <summary>Whether the flag exists in the current cache.</summary>
        public bool Exists => _client.HasFlagInternal(_flagName, _forceRealtime);

        /// <summary>Check if flag is enabled. Delegates to client for metrics tracking.</summary>
        public bool Enabled => _client.IsEnabledInternal(_flagName, _forceRealtime);

        /// <summary>Get variant object. Delegates to client for metrics tracking.</summary>
        public Variant Variant => _client.GetVariantInternal(_flagName, _forceRealtime);

        // Read-only metadata (no metrics needed)
        public ValueType ValueType => _client.GetValueTypeInternal(_flagName, _forceRealtime);
        public int Version => _client.GetVersionInternal(_flagName, _forceRealtime);
        public string Reason => _client.GetReasonInternal(_flagName, _forceRealtime);
        public bool ImpressionData => _client.GetImpressionDataInternal(_flagName, _forceRealtime);
        public EvaluatedFlag Raw => _client.GetRawFlagInternal(_flagName, _forceRealtime);

        #endregion

        #region Variations (pure delegation)

        public string Variation(string fallbackValue)
            => _client.VariationInternal(_flagName, fallbackValue, _forceRealtime);

        public bool BoolVariation(bool fallbackValue)
            => _client.BoolVariationInternal(_flagName, fallbackValue, _forceRealtime);

        public string StringVariation(string fallbackValue)
            => _client.StringVariationInternal(_flagName, fallbackValue, _forceRealtime);

        public int IntVariation(int fallbackValue)
            => _client.IntVariationInternal(_flagName, fallbackValue, _forceRealtime);

        public float FloatVariation(float fallbackValue)
            => _client.FloatVariationInternal(_flagName, fallbackValue, _forceRealtime);

        public double DoubleVariation(double fallbackValue)
            => _client.DoubleVariationInternal(_flagName, fallbackValue, _forceRealtime);

        public Dictionary<string, object> JsonVariation(Dictionary<string, object> fallbackValue)
            => _client.JsonVariationInternal(_flagName, fallbackValue, _forceRealtime);

        #endregion

        #region Variation Details (pure delegation)

        public VariationResult<bool> BoolVariationDetails(bool fallbackValue)
            => _client.BoolVariationDetailsInternal(_flagName, fallbackValue, _forceRealtime);

        public VariationResult<string> StringVariationDetails(string fallbackValue)
            => _client.StringVariationDetailsInternal(_flagName, fallbackValue, _forceRealtime);

        public VariationResult<int> IntVariationDetails(int fallbackValue)
            => _client.IntVariationDetailsInternal(_flagName, fallbackValue, _forceRealtime);

        public VariationResult<float> FloatVariationDetails(float fallbackValue)
            => _client.FloatVariationDetailsInternal(_flagName, fallbackValue, _forceRealtime);

        public VariationResult<double> DoubleVariationDetails(double fallbackValue)
            => _client.DoubleVariationDetailsInternal(_flagName, fallbackValue, _forceRealtime);

        public VariationResult<Dictionary<string, object>> JsonVariationDetails(Dictionary<string, object> fallbackValue)
            => _client.JsonVariationDetailsInternal(_flagName, fallbackValue, _forceRealtime);

        #endregion

        #region OrThrow (pure delegation)

        public bool BoolVariationOrThrow()
            => _client.BoolVariationOrThrowInternal(_flagName, _forceRealtime);

        public string StringVariationOrThrow()
            => _client.StringVariationOrThrowInternal(_flagName, _forceRealtime);

        public int IntVariationOrThrow()
            => _client.IntVariationOrThrowInternal(_flagName, _forceRealtime);

        public float FloatVariationOrThrow()
            => _client.FloatVariationOrThrowInternal(_flagName, _forceRealtime);

        public double DoubleVariationOrThrow()
            => _client.DoubleVariationOrThrowInternal(_flagName, _forceRealtime);

        public Dictionary<string, object> JsonVariationOrThrow()
            => _client.JsonVariationOrThrowInternal(_flagName, _forceRealtime);

        #endregion
    }
}
